velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
   }
 };
 
+template <typename V, typename Compare>
+struct MinMaxByNComplexCompareTypeAccumulator {
+  using TRawValue = typename RawValueExtractor<V>::TRawValue;
+  using BaseType =
+      MinMaxByNAccumulator<V, HashStringAllocator::Position, Compare>;
+  BaseType base;
+  ValueSet valueSet;
+
+  explicit MinMaxByNComplexCompareTypeAccumulator(
+      HashStringAllocator* allocator)
+      : base{allocator}, valueSet{allocator} {}
+
+  ~MinMaxByNComplexCompareTypeAccumulator() {
+    for (auto i = 0; i < base.heapValues.size(); ++i) {
+      freePair(base.heapValues[i]);
+    }
+  }
+
+  int64_t getN() const {
+    return base.n;
+  }
+
+  size_t size() const {
+    return base.size();
+  }
+
+  void checkAndSetN(DecodedVector& decodedN, vector_size_t row) {
+    return base.checkAndSetN(decodedN, row);
+  }
+
+  void compareAndAdd(
+      DecodedVector& decodedCompare,
+      vector_size_t compareIndex,
+      std::optional<V> value,
+      Compare& comparator) {
+    if (base.heapValues.size() < base.n) {
+      auto comparePosition = writeComplex(decodedCompare, compareIndex);
```

Comment:
Let's think about how to extend tests to capture the bug I pointed out earlier. Let's also check other compareAndAdd for similar issues.

Reply:
> This code should not run unconditionally. It should be evaluated only if we are going to store the value in the accumulator.

You mean this above? I think this is not a bug, but to avoid unnecessary operations? (Maybe the comment I quote is not you specified^^)

Reply:
@f0rest9999 Indeed, this issue doesn't affect correctness, but affects performance and in extreme cases may cause query failure due to OOM. We need to ensure both correct and efficient implementation.